### PR TITLE
Change SymIntNode into an intrusive pointer

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82468
* #82555
* #82467
* #82477
* __->__ #82548

This will make the pointer type a single word, which is important
for packing it into an int64_t

This time, this diff doesn't segfault when you build with DEBUG mode; more details at https://github.com/pybind/pybind11/issues/4099

Signed-off-by: Edward Z. Yang <ezyangfb.com>

### CODE REVIEW

File: .github/ci_commit_pins/xla.txt [hunk 1]
```diff
@@ -1 +1 @@
-b3342319e96a0becd139019620d8665605b78475
+1693c1c728c05a3afda946415acc9dfd6adf421d
```
Comment (albanD) [inline .github/ci_commit_pins/xla.txt:1]: Why does this need an xla side change?

File: torch/csrc/utils/python_arg_parser.cpp [hunk 2]
```diff
@@ -335,6 +335,9 @@ auto handle_torch_function_no_python_arg_parser(
       // NOLINTNEXTLINE(clang-diagnostic-writable-strings)
       py::object torch_function =
           PyObject_FastGetAttrString(arg.ptr(), torch_function_name_str);
+      if (!torch_function) {
+        TORCH_INTERNAL_ASSERT(0);
```
Comment (albanD) [inline torch/csrc/utils/python_arg_parser.cpp:339]: nit: TORCH_INTERNAL_ASSERT(torch_function, "???") ?

**Review by albanD — APPROVED** [review]
I have a deja vu effect while looking at this?
Still looks good!

File: .github/ci_commit_pins/xla.txt (hunk 1 above)
Comment (ezyang) [reply .github/ci_commit_pins/xla.txt:1]: XLA uses SymIntNode, so it needs to track the change

**Pull request closed without merging**